ITE/11/README.adoc

```diff
@@ -0,0 +1,278 @@
+= ITE-11: Verifying Attributes in in-toto Attestations
+:source-highlighter: pygments
+:toc: preamble
+:toclevels: 2
+ifdef::env-github[]
+:tip-caption: :bulb:
+:note-caption: :information_source:
+:important-caption: :heavy_exclamation_mark:
+:caution-caption: :fire:
+:warning-caption: :warning:
+endif::[]
+
+.Metadata
+[cols="2"]
+|===
+| ITE
+| 11
+
+| Title
+| Verifying Attributes in in-toto Attestations
+
+| Sponsor
+| link:https://github.com/adityasaky[Aditya Sirish A Yelgundhalli]
+
+| Status
+| Draft
+
+| Type
+| Standards
+
+| Created
+| 2023-01-07
+
+|===
+
+[[abstract]]
+== Abstract
+
+in-toto metadata, whether legacy links or contextual predicates defined under
+the Attestation Framework, can record various attributes about software supply
+chain steps and artifacts. However, layouts do not intrinsically support
+setting constraints on these attributes, passing off these capabilities to the
+"inspections" semantic instead. This ITE proposes adding support for setting
+such constraints in in-toto layouts.
+
+[[specification]]
+== Specification
+
+This ITE proposes embedding attribute verification capabilities within in-toto
+layouts. Specifically, it proposes a new step and inspection specific field
+called `expectedAttributes` that will live alongside the existing
+`expectedMaterials` and `expectedProducts` fields. Unlike those fields,
+`expectedAttributes` will not support artifact rules but will instead evaluate
+each rule using the Common Expression Language (CEL).
+
+If this ITE is applied to the in-toto v1.0 layout schema, each step would look
+as follows.
+
+```yaml
+name: string
+threshold: int
+expectedMaterials: list of artifact rules
+expectedProducts: list of artifact rules
+expectedAttributes: list of CEL expressions
+pubkeys: list of authorized key IDs
+expectedCommand: list of strings
+```
+
+If instead the ITE is applied to ITE-10 layouts with support for the in-toto
+Attestation Framework, the resultant step schema is as follows.
+
+```yaml
+name: string
+command: list of strings
+expectedMaterials: list of artifact rules
+expectedPredicates: list of expectedStepPredicates
+```
+
+Each `expectedStepPredicate` object has the following schema.
+
+```yaml
+predicateType: string
+expectedProducts: list of artifact rules
+expectedAttributes: list of CEL expressions
```

Comment:
FYI, one thing that's done in CEL Policy Templates is to create reusable `terms` for more complicated policies. Here's an example:
https://github.com/google/cel-policy-templates-go/blob/master/test/testdata/binauthz/template.yaml

Not sure if you want to add that proactively, or if you want to wait to see if it's actually a problem.

Reply:
I suspect we'll need something like this eventually but I want to wait and see what shape this takes. In part, this is because I think we need to explore how layout-templates can be written, which may be where we want to support this.

Reply:
Sounds good, feel free to resolve